### ISSUE TITLE
Purpose of FanPool

### ISSUE THREAD

**Issue by thedevop** [mochi-co/mqtt#150]
@mochi-co , can you please elaborate on the purpose of using FanPool? Currently when multiple publishers publish to same or different topics, FanPool will not ensure order nor provide concurrency protection (like PacketID issue).

**Comment by mochi-co** [comment]
Hi @thedevop - sorry for the late reply on this, it's been a very busy week.

You may remember in v1.x.x, packet transmission was handled using a circular ring buffer (https://github.com/mochi-co/mqtt/tree/v1.1.1/server/internal/circ). It was fast and efficient, but it had some downsides which included dropping messages (#54, #109), and unusual race conditions and deadlocks (#101, #107). Additionally, each client required 3 goroutines to support this pattern (the client itself, a reader, and a writer), which was not suitable for everyone under high loads (#80).

However, more significantly, we found that (and this is slightly unrelated to the circular buffer) the dispatch pattern in v1 would cause the whole broke to stall if a connected client was slow in receipt (#95). This was one of the key motivators which forced me to actually get on with finishing v2, and then I had the perhaps naive idea to also upgrade us to MQTT v5 ("I'll just implement mqtt v5 while I'm making this change, how hard could it be").

I still needed some kind of go worker dispatch system, so I had a look at what others were doing. After a fair amount of benchmarking, I concluded that https://github.com/fhmq/hmq/blob/master/pool/fixpool.go was the most performant worker pool option; and in essence our fan pool is almost identical to this (as noted in the fanpool file header). 

During the development of v2, I switched off the fanpool entirely as I needed to monitor the broker internal processes without having to dig through goroutines. It wasn't until pre-release that I re-enabled it, and in doing so found no meaningful performance increase by having it enabled (at least in my environment). However, without the fanpool, we return to issue #95 - if a connected client is slow in receipt, the transmitting client would be delayed in sending messages to all subscribers. Moreover, the client (goroutines) became responsible for issuing messages to subscriptions, which meant publishing a message to a topic with thousands of subscribers would severely limit that client's own throughput, and so the fixed fan pool was re-enabled.

Having read your issue (#149), I am not convinced the fanpool is the correct solution. Neither am I 100% convinced that a goroutine for every worker is suitable either, although it is probably more flexible, and realistically anyone running a broker with hundreds of thousands of clients should probably have more than n*2kb in memory to spare. Especially if we can get clustering developed at some point (I know @bkupidura is working on some promising ideas).

I would be interested to hear your thoughts (and others who might be reading) on this topic, for sure.

My feelings on it:

- I am happy with a go routine per client, although I'm not sure at this stage how it should be structured.
- We could potentially implement the worker pool as an interface and allow developers to use their own implementations if desired.


cc @wind-c @bkupidura as you are both key contributors to this project.

**Comment by thedevop** [comment]
Let's only look at the worst case scenario, as this is addressed, all other cases will be addressed as well. Assuming there is a subscriber to `#` that's very slow at receiving (in fact, you can extend to ALL subscriber are slow at receiving), if the publish rate is > the send rate to that client, no matter what method is employed, at some point this subscriber will lose messages (memory/disk buffer will eventually run out).

If there is a bounded queue, and enqueue is blocking, it will impact all publishers when the queue is full. Currently that queue in set on the publisher side (using FanPoolSize=1):
```
Pub 1 --
       |   -----------------
Pub 2 ---> | FanPool Queue | --> FanPool Worker --> (Msg 1 -> Sub 1, 2, 3...), (Msg 2 -> ...)
       |   -----------------
Pub 3 --
```
As the above illustrates, this doesn't solve [#95](https://github.com/mochi-co/mqtt/issues/95) as when FanPool Queue is full, all publishers will be blocked.

To address that, enqueue has to be non-blocking. If the queue is full, message needs to be dropped. There can be 2 queue implementations:
1. Queue on the publisher side,  like the current FanPool implementation, then entire publish message will be dropped to all subscribers. 
2. Queue on t

**Comment by thedevop** [comment]
Let's only look at the worst case scenario, as this is addressed, all other cases will be addressed as well. Assuming there is a subscriber to `#` that's very slow at receiving (in fact, you can extend to ALL subscriber are slow at receiving), if the publish rate is > the send rate to that client, no matter what method is employed, at some point this subscriber will lose messages (memory/disk buffer will eventually run out).

If there is a bounded queue, and enqueue is blocking, it will impact all publishers when the queue is full. Currently that queue in set on the publisher side (using FanPoolSize=1):
```
Pub 1 --
       |   -----------------
Pub 2 ---> | FanPool Queue | --> FanPool Worker --> (Msg 1 -> Sub 1, 2, 3...), (Msg 2 -> ...)
       |   -----------------
Pub 3 --
```
As the above illustrates, this doesn't solve [#95](https://github.com/mochi-co/mqtt/issues/95) as when FanPool Queue is full, all publishers will be blocked.

To address that, enqueue has to be non-blocking. If the queue is full, message needs to be dropped. There can be 2 queue implementations:
1. Queue on the publisher side,  like the current FanPool implementation, then entire publish message will be dropped to all subscribers. 
2. Queue on the subscriber side, then only message for that subscriber will be dropped
```
                 |-> Sub 1 Queue --> Write Worker -> Sub 1
Pub 1 --> Msg 1 -|-> Sub 2 Queue --> Write Worker -> Sub 2
                 |-> Sub 3 Queue --> Write Worker -> Sub 3
Same for Pub 2, 3...
```
The write worker can be 1:1 (as with v1) to a client (subscriber) or 1:M (like fanpool but based on subscriber instead of publisher), impact radius will be different though.

In summary, the key is to drop messages to slow receiver, there are different implementations that can have different impact radius, the closer the queue is to the subscriber client the smaller the impact.

**Comment by mochi-co** [comment]
@thedevop Apologies again, I am looking and thinking about this, I am just unusually extraordinarily busy 🥶 

I have been considering a somewhat similar approach. I agree that ultimately the solution is that oversaturated clients should be skipped when publishing to subscribers. Having said that, as of MQTT v5, we now have send and receive quotas which ensure that a client cannot send or receive more QOS >0 messages than are allowed. Any messages beyond this number are stored and sent once the quota is no longer zero. 

I wonder if we could expand on this concept somehow to implement an internal queue for each client. One way to do this would be to add a buffered channel for each client (chan Packet). If the channel is at capacity, then the client is skipped. This would give developers the ability to set their dropped message tolerances.

Each client could then have a goroutine which indefinitely drains the channel, writing these messages to the client. 

Another alternative would be to somehow have a fixed worker queue which operates on these different channels, so as not to have an additional goroutine per client - however, I've not yet determined a simple mechanism to do this which wouldn't eventually run into the issues above.

What do you think?

**Comment by thedevop** [comment]
No worries @mochi-co, totally understand.

Each client with its own fixed (configurable?) buffered channel and write worker is cleanest, and has the least impact radius with a slow receiving subscriber.

**Comment by thedevop** [comment]
Addressed in [PR 165](https://github.com/mochi-co/mqtt/pull/165)
